tests/jenkins/jenkinsjob-regression-files/data-prepper/distribution-artifacts.jenkinsfile.txt

```diff
@@ -11,9 +11,27 @@
          distribution-artifacts.stage(Sign Artifacts, groovy.lang.Closure)
             distribution-artifacts.script(groovy.lang.Closure)
                distribution-artifacts.signArtifacts({artifactPath=/tmp/workspace/release/archives/linux/build/distributions, signatureType=.sig, distributionPlatform=linux})
+                  signArtifacts.fileExists(/tmp/workspace/sign.sh)
```

Comment:
I'd rather we not have this text pasted here. 

1. These details don't matter to the author of the Data Prepper pipeline.
2. If you change the implementation fo `signArtifacts`, this test will fail.
3. This will likely become boilerplate. In my experience boilerplate is just copied and pasted mostly blindly. Again, the author of this pipeline doesn't care that that this users `usernamePassword` credentials to access S3.

Reply:
A difference of opinion -
I think this test should fail with changes to `signArtifacts `. It shows that this piece of code is being affected by the change. If we mock the `signArtifacts` function, it would only hide the dependency of this code to the functions making weakening tests. We should avoid mocking wherever possible.

Also, failing these tests would make the developer aware of the blast radius of the change.

Reply:
I don't think the tests are weakened. 

We are verifying the following:
* The `TestSignArtifacts` test should verify that it works. This should be the only test that needs to verify the implementation.
* Each possible job test verifies its usage of `TestSignArtifacts`.

Thus, we know signArtifacts is correct and is used correctly.

Reply:
I think part of what is missing is that `TestSignArtifacts` is currently not a fully robust test suite. Improving this would help.

For example, if you change this line:
https://github.com/opensearch-project/opensearch-build/blob/main/vars/signArtifacts.groovy#L27

to:

```
$WORKSPACE/sign.sh ${args.artifactPath} --sigtype=.sig --component=${args.component} --type=${args.type}
```

The tests still pass. 

It is because there is one test and it tests using only `.sig`. You can see this here: https://github.com/opensearch-project/opensearch-build/blob/main/tests/jenkins/TestSignArtifacts.groovy#L24

I suggest improving the tests for `TestSignArtifacts` to ensure that `signArtifacts` is correct.

Reply:
In my opinion - Mocking code is never a good idea, we should avoid it as much as we can. No test is perfect and there may always be cases where something might fail. 100% coverage can till have code failures.

Coming to the case here, our jenkins test are not bulletproof, but we are getting there as we improve the process and build the infrastructure around it. Thanks a lot for contributing #1483, I added verifying params here.

Reply:
The Jenkins JUnit library is a mocking framework. All of these tests use mocking.

This is still "magical text" and I expect that will be used blindly.

Reply:
We use mocking at places which we cannot execute with the current framework. If could execute the bash, we should.